src/vfs2.c

```diff
@@ -1202,6 +1190,100 @@ static sqlite3_int64 wal_offset_from_cursor(uint32_t page_size, uint32_t cursor)
 		    (sqlite3_int64)page_size);
 }
 
+/**
+ * Read the given WAL file from beginning to end, initializing the WAL-index in
+ * the process.
+ *
+ * The process stops when we reach a frame whose checksums are invalid, or
+ * after the last valid commit frame, or on the first unsuccessful read, or at
+ * the end of the transaction designated by `stop`, if that argument is
+ * non-NULL.
+ *
+ * On return, the WAL-index header is initialized with mxFrame = 0 and other
+ * data matching the given WAL, and the page numbers for all frames up to the
+ * stopping point are recorded in the WAL-index; the hash tables are not
+ * initialized.
+ *
+ * Returns the stopping point in units of frames, which becomes the wal_cursor.
+ */
+static int walk_wal(sqlite3_file *wal,
+		    sqlite3_int64 size,
+		    struct wal_hdr hdr,
+		    const struct vfs2_wal_slice *stop,
+		    uint32_t *wal_cursor,
+		    struct shm *shm)
+{
+	uint32_t page_size = ByteGetBe32(hdr.page_size);
+	struct cksums sums = { ByteGetBe32(hdr.cksum1),
+			       ByteGetBe32(hdr.cksum2) };
+	int rv;
+
+	/* Check whether we have been provided a stopping point that corresponds
+	 * to a transaction in the current WAL. (It's possible that the stopping
+	 * point corresponds to a transaction that's in WAL-prev instead.) */
+	bool have_stop = stop != NULL && salts_equal(stop->salts, hdr.salts);
```

Comment:
Good catch, this needs work. If the stopping point is at the end of WAL-prev then this is benign, it's equivalent to stopping before the first frame of WAL-cur. If the stopping point is located in WAL-prev somewhere before the end, then that indicates a bug (because everything in WAL-prev should be committed). And of course, if the salts for the stopping point don't match either of the WALs, something has gone wrong. To implement this logic fully we need to walk WAL-prev so we can determine where it ends.

Reply:
Nice catch about the distinction between end of WAL-prev and somewhere in the middle.

Maybe we can implement this as a bugfix in a separate PR to avoid this one taking longer to merge, what do you think?

Reply:
Agree, I'll just add a TODO.